[SecuritySolution] Add timeline save button / remove timeline's autosave

janmonschke opened the pull request:
## Summary

(associated issue: https://github.com/elastic/kibana/issues/165053)

In this PR we're adding a dedicated save button to the timeline flyout and at the same time remove the autosave behaviour of timeline. This should mitigate previous issues where in multi-tenant environments, timelines were overridden by accident. This change also allows threat hunters to draft changes of a timeline and discard them when they're not needed anymore (simply by not pressing `save`).

**Flow for unsaved timeline:**


https://github.com/elastic/kibana/assets/68591/172ab3dc-c4a6-448a-8008-241436d9432c



**Flow for previously saved timeline:**



https://github.com/elastic/kibana/assets/68591/03e6f626-7693-4ff3-a55f-1b970a371370



In case the user cannot edit timelines, it will show the same tooltip that was shown before for the timeline edit buttons:

<img width="445" alt="Screenshot 2023-10-18 at 09 22 30" src="https://github.com/elastic/kibana/assets/68591/cb72c26a-3639-4833-afd6-c352ef6c7792">

### Architecture

In order to turn off the autosave feature and add highlighting of changes, the timeline epics needed to be changed. Let's look at how they were working before and after:

#### Before
```mermaid
flowchart TB
    A(Action);

    A --> epic_timeline;
    A --> epic_note;
    A --> epic_pinned_event;
    A --> epic_favorite;

    subgraph epic_timeline
        B(createTimelineEpic)
        B1(filterAction)
        B2(persistTimeline)
        B-->B1
        B1--persistQueue.next-->C
        C(dispatcherTimelinePersistQueue)
        C-->X(delay 500ms)
    end

    subgraph epic_note
    D(createTimelineNoteEpic)
    D1(filterAction)
    D2(epicPersistNote)
    D3(persistNote)

    D-->D1
    D1--persistQueue.next-->C
    D2-->D3
    end

    subgraph epic_pinned_event
    E(createTimelinePinnedEventEpic)
    E1(filterAction)
    E2(epicPersistPinnedEvent)
    E3(persistPinnedEvent)

    E-->E1
    E1--persistQueue.next-->C
    E2-->E3
    end

    subgraph epic_favorite
    F(createTimelineFavoriteEpic)
    F1(filterAction)
    F2(epicPersistTimelineFavorite)
    F3(persistFavorite)
    F-->F1
    F1--persistQueue.next-->C
    F2-->F3
    end

    subgraph so_saving_pipe
        direction TB
        startTimelineSaving-->updateTimeline-->updateSO-->endTimelineSaving-->takeUntilWithSideffects
    end

    X-->B2
    X-->D2
    X-->E2
    X-->F2

    B2--embedded-->so_saving_pipe
    D3--embedded-->so_saving_pipe
    E3--embedded-->so_saving_pipe
    F3--embedded-->so_saving_pipe
```

Timeline and it's associated SOs have their own epics that all filter for relevant actions. Eventually, when they see a matching action, they will trigger the persistence queue (which, for some reason, is piped together with the timeline epic 🤷 ).

E.g. a `pinEvent` action will trigger the persistence queue and in turn start the `persistPinnedEvent` epic. That logic lives in the embedded persistence queue. The `persistPinnedEvent` epic has an embedded pipe that will pretend the timeline is saving by dispatching `startTimelineSaving`, then do its own persistence calls and end on `endTimelineSaving`. This will show loading indicators in the UI. The timeline model itself, is not saved in this case. That only happens through the timeline epic.

The timeline epic filters for a (longer) list of actions that will trigger timeline persistence. It does some filtering for the action and has some (global) side effects that are required by all the other epics as well ([see here](https://github.com/elastic/kibana/blob/761fa5173c22dcca20ee4a6e499d85237f8773bc/x-pack/plugins/security_solution/public/timelines/store/timeline/epic.ts#L172-L183)). It then starts its own pipe that includes `startTimelineSaving` and `endTimelineSaving` and a call to persist the timeline.

Neither of these epics are (unit-)tested, so making changes to these epics would lower the confidence that everything was well implemented. Therefore, the change in architecture introduces a new, parallel epic that barely touches the timeline epic and has no impact on the other related epics.

#### After

```mermaid
flowchart TB
    A(Action);

    A --> epic_timeline;
    A --> epic_changed;
    A --> epic_note_pinned_event_favorite;

    subgraph epic_timeline
        B(createTimelineEpic)
        B1(isSaveTimeline)
        B2(persistTimeline)
        B-->B1
        B1--persistQueue.next-->C
        C(dispatcherTimelinePersistQueue)
        C-->X(immediateNoDelay)
    end

    subgraph epic_note_pinned_event_favorite
    D(createSOEpic)
    D1(filterAction)
    D2(epicPersistSO)
    D3(persistSO)

    D-->D1
    D1--persistQueue.next-->C
    D2-->D3
    end

    subgraph so_saving_pipe
        direction TB
        startTimelineSaving-->updateTimeline-->updateSO-->endTimelineSaving-->takeUntilWithSideffects
    end

    subgraph epic_changed
        direction TB
        createChangedEpic --> filterForChanges --> mapToChangedAction
    end

    X-->B2
    X-->D2
    

    B2--embedded-->so_saving_pipe
    D3--embedded-->so_saving_pipe
```

(note: the note, pinned event and favorite epics remain unchanged, I just merged them into one in this diagram to simplify it and highlight the changed epic)

The newly introduce `changed` epic uses the list of actions that would previously trigger the `timeline` epic persistence. All actions in that list are then mapped to `timelineChanged` actions which in turn will then set the `changed` prop on timeline to `true`. Timeline persistence is now only triggered by the `saveTimeline` action.

### UI changes

- The `EditTimelineButton` has been removed, and a new `SaveTimelineButton` has been introduced. This button, as seen in the videos above, sits at the top of the timeline flyout. Whenever the save button is pressed, it will open up the `SaveTimelineModal` which is the former `TimelineEditModal` with minimal changes.

- The `TimelineStatusInfo` component has been adapted to show when the timeline has unsaved changes.

- A new component (`TimelineSavePrompt`) was introduced. It takes care of the case we want to show the `SaveTimelineModal` when navigating away to another solution with an changed / unsaved timeline. This logic was previously embedded/hidden/stuffed into the `EditTimelineButton`

https://github.com/elastic/kibana/blob/884bcb262514c9253a8df597a88a4950ad13397c/x-pack/plugins/security_solution/public/timelines/components/timeline/header/edit_timeline_button.tsx#L76

(the logic isn't documented and is brittle and it can break easily during a refactor)

The dedicated `TimelineSavePrompt` cleans up that logic and passes the cypress tests that were testing this scenario. (https://github.com/elastic/kibana/pull/166957/commits/76e45dad54dae817a246282afae1d0b8a3316024)

- A `refreshedAt` prop was introduced to the timeline model that indicates when the timeline events were last refreshed. Previously this was tied to when the timeline was last changed. However, a refresh of the data does not necessarily trigger a change event (in case `from` and `to` stay the same).

- The new `changed` prop is also taken into account when showing the timeline save prompt when navigating away from security solution with unsaved changed.

- There were many cypress tests that relied on the the autosave feature of timeline. These tests now use the new `saveTimeline` task. In turn `waitForTimelineChanges` has been removed, as it's not accurate enough anymore.

### Checklist
- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios

janmonschke committed:
Add timeline save button

janmonschke committed:
properly distinguish between the timeline save prompt and the other modals

janmonschke committed:
autosaved -> saved

janmonschke committed:
remove unneeded `tooltip` prop

janmonschke committed:
add save button tests

kibanamachine committed:
Merge branch 'main' into save-timeline-improvements-2

janmonschke committed:
Merge branch 'main' into save-timeline-improvements-2

janmonschke committed:
Merge branch 'main' into save-timeline-improvements-2

janmonschke committed:
Move save button to top

- Also removes the description
- Removes the edit buttons from title and description
- The timeline status info now shows as a tooltip on the save button
- The timeline edit button has been removed from NotePreviews

janmonschke committed:
Merge branch 'main' into save-timeline-improvements-2

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/eslint --no-cache --fix'

janmonschke committed:
remove unused translation labels

janmonschke committed:
fix save timeline button tests

janmonschke committed:
fix cypress tests

janmonschke committed:
Merge branch 'main' into save-timeline-improvements-2

janmonschke committed:
remove unnecessarily dispatched actions

The time range doesn't change in `onRefresh`, therefore the time in redux doesn't need updating.

This prevents unnecessary store updates and a timeline save.

janmonschke committed:
Revert "remove unnecessarily dispatched actions"

This reverts commit 51250abb6b23210663bc9d9a5be5a3fb97efd876.

janmonschke committed:
Don't update the redux time range if the time range didn't actually change

janmonschke committed:
Revert "Don't update the redux time range if the time range didn't actually change"

This reverts commit 39a3c2af941d43b02b2e4181801b2c37a1a508f5.

janmonschke committed:
introduce `changed` prop

- timeline does not autosave anymore
- a set of actions has been defined that will mark the timeline as changed (in a new epic)
- only saveTimeline() actions will start the persistence
- makes sure to respect `changed` when showing the save prompt for unsaved timelines upon navigation

janmonschke committed:
updatedAt -> refreshedAt

updatedAt, which came from `updated`, stood for when the timeline was last saved. In this case we are actually interested in when the timeline (query) was last refreshed.

janmonschke committed:
Merge branch 'main' into save-timeline-improvements-2

janmonschke committed:
fix row-renderer cypress tests

janmonschke committed:
fix correlation tab tests

janmonschke committed:
fix unsaved timeline tests

janmonschke committed:
fix dataprovider cypress tests

janmonschke committed:
fix search or filter cypress tests

janmonschke committed:
fix timeline template cypress tests

janmonschke committed:
make dataprovider tests mergeable

janmonschke committed:
Merge branch 'main' into save-timeline-improvements-2

janmonschke committed:
fix helpers tests

janmonschke committed:
fix use_create_timeline tests

janmonschke committed:
show green button when timeline has unsaved changes

janmonschke committed:
fix flaky correlation tab tests

janmonschke committed:
automatically save the timeline when a saved search id is created

janmonschke committed:
allow for save actions when one is in flight timeline has changed

janmonschke committed:
since autosave is gone, we don't need to delay the queue anymore

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/precommit_hook.js --ref HEAD~1..HEAD --fix'

janmonschke committed:
fix url state cypress tests

janmonschke committed:
Merge branch 'main' into save-timeline-improvements-2

janmonschke committed:
move use_edit_timeline_modal logic into save timeline button

janmonschke committed:
waitForTimelineChanges() is not needed anymore

janmonschke committed:
"Name timeline" doesn't make sense anymore

janmonschke committed:
remove unused import

janmonschke committed:
change all action filters to object lookups

Improves filter performance from O(n) to O(1)

janmonschke committed:
move away from deprecated `empty()`

janmonschke committed:
use sets instead of objects

- no need to keep values around
- lookup might be even faster :D

janmonschke committed:
make autofocus smarter

janmonschke committed:
refreshing data does not update the range

janmonschke committed:
add missing`hasRangeChanged`

janmonschke commented on x-pack/plugins/security_solution/public/common/store/data_table/epic_local_storage.ts:
```diff
@@ -34,7 +34,7 @@ const {
 export const isNotNull = <T>(value: T | null): value is T => value !== null;
 
-const tableActionTypes = [
+const tableActionTypes = new Set([
```
The reason I changed all the epic filter action checks to using Sets is that they are saving a hundreds of unnecessary iterations. From reload to displaying a timeline, roughly 50 actions are dispatched.

Each epic receives all actions, so just the local-storage epic and the timeline-changed epic alone would in a worst case scenario need 1650 iterations when using array.includes(). Using Set, it's 100 in a worst case scenario.

janmonschke replied:
Sure thing, let me elaborate 👍  My point has to do with the amount of iterations that each epic's filter method has to do. So I counted how many redux actions are fired from the moment when the site loads until a timeline is show. That number is roughly 50.

The `local-storage-epic` and the `timeline-changed` epic (or previously the regular `epic.ts`) have a list of 13 and 20 actions that they respond to. For each of the 50 actions that are dispatched, these two epics then create 50 * 13 or 50 * 20 iterations to find elements in an array. That makes 1650 calls only to show a timeline.

Using a `Set`, which looks up elements in constant time (vs `Array`'s linear time), that lookup happens in 50 and 50 iterations because the set doesn't need to traverse its list of elements.

logeekal replied:
Got it. thanks for the explanation. It is a great catch.

janmonschke commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into save-timeline-improvements-2

janmonschke committed:
address flakyness of some timeline cypress tests

janmonschke commented:
@elasticmachine merge upstream

janmonschke commented:
@elasticmachine merge upstream

janmonschke commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into save-timeline-improvements-2

logeekal commented on x-pack/plugins/security_solution/public/timelines/components/timeline/query_bar/index.tsx:
```diff
@@ -221,6 +221,7 @@ export const QueryBarTimeline = memo<QueryBarTimelineComponentProps>(
             isInvalid: false,
             isQuickSelection,
             timelineId,
+            hasRangeChanged: true,
```
I am not very clear with the introduction of this new variable. Could you please provide me a small explanation of what is it doing?

janmonschke replied:
`dispatchUpdateReduxTime` is dispatching A LOT of actions and some of them, in this case the `updateRange` actions, don't need to fire when the time range hasn't change. Changing the range leads to a change in the timeline model, so that should only happen when the range has actually changed. Otherwise, the timeline shows as `changed` when it's loaded.

janmonschke replied:
@logeekal Added a comment here https://github.com/elastic/kibana/pull/169239/commits/c987abaf2458e0ad4129bab981382d827a95cb62

logeekal replied:
Got it.. Thanks.

logeekal commented on x-pack/plugins/security_solution/public/timelines/store/timeline/epic.ts:
```diff
@@ -182,8 +132,9 @@ export const createTimelineEpic =
-            !timelineObj.isLoading &&
+            action.type === saveTimeline.type &&
+            // don't save again when already saving, unless the timeline changed
+            (!timelineObj.isLoading || timelineObj.changed) &&
```
Shouldn't it be 

```javascript
!timelineObj.isSaving
```

janmonschke replied:
I think my comment was misleading here. I fixed it in https://github.com/elastic/kibana/pull/169239/commits/8df3e7fa9df271be296b7a77cbde24b26043443c.
@logeekal let me know if it makes more sense.

`isLoading` isn't really tested and so I did not want to change this logic too much to.

logeekal reviewed (commented):
Below is my code-review and mostly it is looking very good 🚀 . I have added few small questions/comments 👇 .

Now I will do desk testing and will revert back. Hopefully by tomorrow. 

Additionally, there was a comment from @paulewing that every click to `Save` button should be treated as `Save a New` action very similar to `discover`. If user types a new name, it is a `create` operation for new timeline, otherwise it is `update` operation for existing timeline.

I am not sure if you and paul already discussed about  it.


Also, right now I am not approving it because @michaelolo24 said that we need to ask QA to test this before we merge it into main.

janmonschke commented:
> Additionally, there was a comment from @paulewing that every click to `Save` button should be treated as `Save a New` action very similar to `discover`. If user types a new name, it is a `create` operation for new timeline, otherwise it is `update` operation for existing timeline.

@logeekal This will come in a second iteration through a new form element in the save modal, yep 👍

janmonschke commented on x-pack/plugins/security_solution/public/timelines/components/open_timeline/note_previews/index.tsx:
```diff
@@ -227,7 +226,7 @@ export const NotePreviews = React.memo<NotePreviewsProps>(
                     size="l"
                   />
                 ),
-                actions: <EditTimelineButton timelineId={timelineId} initialFocus="description" />,
```
This will only remove the edit button from the "first" note which is the description. It was a request from product and design to remove this button as well. It's mentioned in https://github.com/elastic/kibana/issues/165053 but further down :)

logeekal replied:
Thanks for explanation @janmonschke

janmonschke committed:
dont' render ReadMoreButton when there is no description

janmonschke commented on x-pack/plugins/security_solution/public/timelines/components/timeline/header/save_timeline_button.tsx:
```diff
@@ -0,0 +1,106 @@
+      <>
+        <EuiButton
+          fill
+          color={urgeToSave ? 'success' : 'primary'}
```
@logeekal That's done by the `disabled` prop 👍

michaelolo24 replied:
Made a comment about this, but feel free to follow up with @paulewing. I believe we want to keep the color 'primary' no matter the state, but please verify with him

janmonschke committed:
Commenting `hasRangeChanged`

janmonschke commented on x-pack/plugins/security_solution/public/timelines/components/timeline/header/save_timeline_button.test.tsx:
```diff
@@ -0,0 +1,125 @@
+      fireEvent.click(saveTimelineBtn);
+
+      await waitFor(() => {
+        expect(screen.getByTestId(TEST_ID.SAVE_TIMELINE_MODAL)).toBeInTheDocument();
```
Good point, will add that!

janmonschke replied:
@logeekal The color change and the tooltip behaviour is not observable from jest since tooltips don't work there properly. I added a Cypress test instead: https://github.com/elastic/kibana/pull/169239/commits/22b3c7050fa1bca1a00a5041313215d545f7155b 🎉

logeekal replied:
Okay.. Cool. Thank you.

janmonschke committed:
comment isLoading || changed logic

janmonschke commented on x-pack/plugins/security_solution/public/timelines/components/flyout/header/index.tsx:
```diff
@@ -286,63 +279,25 @@ const TimelineDescriptionComponent: React.FC<FlyoutHeaderProps> = ({ timelineId
   return (
     <EuiText size="s" data-test-subj="timeline-description">
-      <ReadMoreButton description={description || commonI18n.DESCRIPTION} onclick={onReadMore} />
+      <ReadMoreButton description={description || ''} onclick={onReadMore} />
```
@logeekal That was there to please TypeScript but you're right, it doesn't look good. I fixed it here: https://github.com/elastic/kibana/pull/169239/commits/6ac51ab59719f1a0bd17521f47e6dab3c4b99cd2

janmonschke committed:
add tooltip tests

logeekal commented:
@janmonschke 

I desk tested PR today and expect 1 scenario all are working well.

 When I save a timeline and then try to save it again by changing its name, button always remains Green irrespective how many times you try to `Save` the timeline. See the demonstration below: 

<video src="https://github.com/elastic/kibana/assets/7485038/4eb3b6f1-0039-491f-bf33-876809323f57" />

janmonschke committed:
fix flakiness around `changed`

janmonschke committed:
prevent undefined values in pipes, they can stop the pipes

janmonschke committed:
no need to debounce anymore

janmonschke committed:
take events forerver

janmonschke committed:
remove unused import

janmonschke commented:
> When I save a timeline and then try to save it again by changing its name, button always remains Green irrespective how many times you try to `Save` the timeline. 

@logeekal Nice find. This issue appears to have been there before with autosave. It seems that the issue was that with lots of undefined values, observables stopped and so some actions wouldn't actually fire anymore: https://github.com/elastic/kibana/pull/169239/commits/8152b5e44d11bd9ea71bd5e4a10ba78954048573

Let me know if it fixed it for you :)

janmonschke commented:
@elasticmachine merge upstream

janmonschke commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into save-timeline-improvements-2

michaelolo24 commented:
Sorry, just a few comments and a couple points of feedback I forgot to mention from a convo with paul

1. Can we bring back the `unsaved` text next to the title when it's not saved?
2. Can the save button color stay the same blue color even in the unsaved state?

This might be related to the one @logeekal  found, but one bug I saw when testing was that saving an already saved timeline and then changing the name leaves the timeline as unsaved. The api returns the below error:

<img width="626" alt="image" src="https://github.com/elastic/kibana/assets/17211684/8532036f-fc25-45c8-8787-b97ff9d9cb41">

This probably never presented as users had autosave enabled. A video is below (The issue resolves after refreshing the page)

https://github.com/elastic/kibana/assets/17211684/7323b629-8179-4b41-a843-1aebbabe667b

michaelolo24 commented on x-pack/plugins/security_solution/public/common/hooks/timeline/use_timeline_save_prompt.ts:
```diff
@@ -102,15 +103,15 @@ export const useTimelineSavePrompt = (
         !(nextAppId ?? '').includes(APP_ID) &&
-        timelineStatus === TimelineStatus.draft &&
-        updated != null
+        (changed || (timelineStatus === TimelineStatus.draft && updated != null))
```
👌🏾

michaelolo24 commented on x-pack/plugins/security_solution/public/timelines/components/flyout/header/index.tsx:
```diff
@@ -286,63 +279,26 @@ const TimelineDescriptionComponent: React.FC<FlyoutHeaderProps> = ({ timelineId
-    },
-    [dispatch, timelineId]
-  );
+  const onReadMore = useCallback(() => {
```
Lol, just found that this existed. Never had a timeline description long enough 😅

michaelolo24 commented on x-pack/plugins/security_solution/public/timelines/components/flyout/header/index.tsx:
```diff
@@ -286,63 +279,26 @@ const TimelineDescriptionComponent: React.FC<FlyoutHeaderProps> = ({ timelineId
 
 const TimelineDescription = React.memo(TimelineDescriptionComponent);
 
-const TimelineStatusInfoComponent: React.FC<FlyoutHeaderProps> = ({ timelineId }) => {
```
Yea, can we keep this component and just use the text `saved` and if there are `changes` have the text 'Unsaved Changes' or something similar?

janmonschke replied:
That's a great idea. Right now it's shown in the button tooltip which isn't the best place imo. Plus, having it only visible on hover, causes all sorts of issues in unit testing and tends to be flaky in integration testing.

michaelolo24 commented on x-pack/test/security_solution_cypress/cypress/tasks/common.ts:
```diff
@@ -25,15 +25,22 @@ export const API_AUTH = Object.freeze({
-export const API_HEADERS = Object.freeze({ 'kbn-xsrf': 'cypress' });
+export const API_HEADERS = Object.freeze({
+  'kbn-xsrf': 'cypress',
+  [ELASTIC_HTTP_VERSION_HEADER]: [INITIAL_REST_VERSION],
```
🔥 , thanks for adding this!

michaelolo24 commented on x-pack/test/security_solution_cypress/cypress/tasks/timeline.ts:
```diff
@@ -344,6 +346,29 @@ export const expandFirstTimelineEventDetails = () => {
   cy.get(TOGGLE_TIMELINE_EXPAND_EVENT).first().click({ force: true });
 };
 
+export const saveTimeline = () => {
```
This should make timeline tests much more reliable. Thanks for adding! 👌🏾

michaelolo24 commented on x-pack/plugins/security_solution/public/timelines/store/timeline/reducer.ts:
```diff
@@ -552,4 +542,14 @@ export const timelineReducer = reducerWithInitialState(initialTimelineState)
       },
     },
   }))
+  .case(setChanged, (state, { id, changed }) => ({
```
simple and clean 👌🏾

michaelolo24 commented on x-pack/plugins/security_solution/public/timelines/store/timeline/epic_changed.ts:
```diff
@@ -0,0 +1,80 @@
+  updateKqlMode.type,
+  updateProviders.type,
+  updateTitleAndDescription.type,
+
```
@logeekal should we add the logic for es|ql here as well?

logeekal replied:
@michaelolo24 , do you mean when `savedSearchId` is updated? I would agree that we can add it here..

Currently, @janmonschke  has put in `saveTimeline` logic in the ESQL/Discover update saved search module itself.

What do you think @janmonschke ?

michaelolo24 commented on x-pack/plugins/security_solution/public/timelines/store/timeline/epic.ts:
```diff
@@ -182,21 +132,20 @@ export const createTimelineEpic =
-            !timelineObj.isLoading &&
+            action.type === saveTimeline.type &&
+            // don't save when the timeline (or its related queries are loading), unless the timeline changed
+            (!timelineObj.isLoading || timelineObj.changed) &&
```
Yea, this might be the scenario that you run into where you are unable to save the timeline while data is loading. We actually should change this to `isSaving` because there's no need to wait for data to load. It would fix an existing bug, but it's hard to reproduce unless your in an environment under load. We can do it in a separate PR though if that makes you more comfortable

janmonschke replied:
Yep this caused all sorts of issues: https://github.com/elastic/kibana/pull/169239/commits/82f741a78a12b382c93854c854ee8fd4a6e9340b

logeekal commented:
>Nice find. This issue appears to have been there before with autosave. It seems that the issue was that with lots of undefined values, observables stopped and so some actions wouldn't actually fire anymore: https://github.com/elastic/kibana/commit/8152b5e44d11bd9ea71bd5e4a10ba78954048573

Honestly, looking this commit in the isolation, I am not sure what is going wrong. So I desk tested it again today, it seems button color is correct as of now as @michaelolo24 has pointed out, timeline is not being actually saved because of below response.

```json
{
    "message": "Timeline template version conflict. The provided templateTimelineVersion does not match the current template.",
    "status_code": 409
}
```

logeekal commented:
>Sorry, just a few comments and a couple points of feedback I forgot to mention from a convo with paul
>
> Can we bring back the unsaved text next to the title when it's not saved?

@michaelolo24, @janmonschke  , What do you think about the badge changes by discover [regarding pending changes](https://github.com/elastic/kibana/pull/169548)?

I think it does not look very beautiful but it is functional..

janmonschke committed:
fix: timeline template and timeline version conflict were using the same error message

janmonschke committed:
remove AutoSaveWarning

There are several reasons for removing this component:

- it wasn't actually working anymore. The code that should update the autosave message was never reached because the server returns no timeline response in the case of a version conflict, it seems to have used to do that.
- what this component tried to do was to "refresh" the local timeline with the remote one. However, this does not necessarily solve the problem because:
  - it will still overwrite local changes
  - since the response, the other party changing the same timeline might have updated the timeline again

Therefore the component was removed now and replaced with a generic error message, informing the user that they'll have to refresh the page (and lose their changes).

This is not ideal, but there is no bullet-proof way to solve this version conflict without real-time synchronisation.

janmonschke commented:
> > Nice find. This issue appears to have been there before with autosave. It seems that the issue was that with lots of undefined values, observables stopped and so some actions wouldn't actually fire anymore: [8152b5e](https://github.com/elastic/kibana/commit/8152b5e44d11bd9ea71bd5e4a10ba78954048573)
> 
> Honestly, looking this commit in the isolation, I am not sure what is going wrong. So I desk tested it again today, it seems button color is correct as of now but as @michaelolo24 has pointed out, timeline is not being actually saved because of below response.
> 
> ```json
> {
>     "message": "Timeline template version conflict. The provided templateTimelineVersion does not match the current template.",
>     "status_code": 409
> }
> ```

@logeekal @michaelolo24 

Thanks for having a look, you two. There seems to indeed be a problem with a race condition somewhere now. And I think I narrowed it down to the saved-search-id flow. When I disable that flow, I don't get conflict errors anymore. However, I think it has to do with the "global" nature of some of the timeline epic parts, rather then the saved-search-id flow. I think it makes sense to integrate the saved search logic into the timeline epic.

Have a look at the diagram proposing an alternative flow:

```mermaid
flowchart TD
    A(Current)
    B(Proposed)

    A-->|spinner starts|A1(TIMELINE_SAVE_START)
    A1-->A2(TIMELINE_SAVE)
    A2-->|spinner ends|A3(TIMELINE_SAVE_END)
    A3-->|delay- spinner starts|A4(SAVE_SAVED_SEARCH)
    A4-->A5(TIMELINE_SAVE_START)
    A5-->A6(TIMELINE_SAVE)
    A6-->|spinner ends|A7(TIMELINE_SAVE_END)

    B-->|spinner starts|B1(TIMELINE_SAVE_START)
    B1-->B2(TIMELINE_SAVE)
    B2-->B3(SAVE_SAVED_SEARCH)
    B3-->B4(TIMELINE_SAVE)
    B4-->|spinner ends|B5(TIMELINE_SAVE_END)
```

In the current flow, the save modal closes when timeline was saved but before the saved search id is saved, the loading spinner does not continuously spin, elements are disabled/enabled multiple times.

In the proposal, the save modal would not close until the second `TIMELINE_SAVE` has finished. The user will not be able to make changes while the save is happening in the background.

On a related note, that error message is misleading, since it's mentioning a conflict for the `timelineTemplateVersion` but we're not using timeline templates. I fount out that both conflicts use the same error message, so I fixed it in https://github.com/elastic/kibana/pull/169239/commits/e0c8ed98ec0001a71e88e4f964ed40a0d3778c86.

Then I saw that there is supposed to be a message + refresh component showing on timeline conflicts. However, that code hasn't been working for some time and I removed it. The reasons can be found in the commit message for https://github.com/elastic/kibana/pull/169239/commits/27724c0a2915e2691db54637b7d64fe087289851. For convenience, I'll paste them here as well:

> There are several reasons for removing this component:
> - it wasn't actually working anymore. The code that should update the autosave message was never reached because > the server returns no timeline response in the case of a version conflict, it seems to have used to do that.
> - what this component tried to do was to "refresh" the local timeline with the remote one. However, this does not necessarily solve the problem because:
>  - it will still overwrite local changes
>  - since the response, the other party changing the same timeline might have updated the timeline again
>
> Therefore the component was removed now and replaced with a generic error message, informing the user that they'll have to refresh the page (and lose their changes).
>
> This is not ideal, but there is no bullet-proof way to solve this version conflict without real-time synchronisation.

logeekal commented:
>Thanks for having a look, you two. There seems to indeed be a problem with a race condition somewhere now. And I think I narrowed it down to the saved-search-id flow. When I disable that flow, I don't get conflict errors anymore. However, I think it has to do with the "global" nature of some of the timeline epic parts, rather then the saved-search-id flow. I think it makes sense to integrate the saved search logic into the timeline epic.
>
>Have a look at the diagram proposing an alternative flow:

@janmonschke , I think your proposed approach is better from UX perspective as well. since, current approach leads to `not great` UX because of multiple time loading indicator starts and stops. 

If it also resolves 409 status code. I like it 🚀 .

janmonschke committed:
remove flakiness/conflicts when saving timeline and surface errors

janmonschke commented:
@logeekal With the changes/fixes introduced in https://github.com/elastic/kibana/pull/169239/commits/82f741a78a12b382c93854c854ee8fd4a6e9340b, we don't actually need to change the flow of actions. So maybe we should just stick with the current approach and clean it up in a follow-up PR. Wdyt?

logeekal commented:
> @logeekal With the changes/fixes introduced in https://github.com/elastic/kibana/commit/82f741a78a12b382c93854c854ee8fd4a6e9340b, we don't actually need to change the flow of actions. So maybe we should just stick with the current approach and clean it up in a follow-up PR. Wdyt?

I agree. That seems to have fixed the issue. I will do one more round of testing tomorrow.

janmonschke committed:
EMPTY is not a valid action, so we're working with a filter function instead

janmonschke commented:
@michaelolo24 thanks for the review :) 

>     1. Can we bring back the `unsaved` text next to the title when it's not saved?

Yep, great idea.

>     2. Can the save button color stay the same blue color even in the unsaved state?

Sure, I can change that back.

> This might be related to the one @logeekal found, but one bug I saw when testing was that saving an already saved timeline and then changing the name leaves the timeline as unsaved. The api returns the below error:
> <img alt="image" width="626" src="https://user-images.githubusercontent.com/17211684/278138817-8532036f-fc25-45c8-8787-b97ff9d9cb41.png">

This behaviour should be fixed now 🎉

janmonschke committed:
bring back timeline status

janmonschke commented:
> @michaelolo24, @janmonschke , What do you think about the badge changes by discover [regarding pending changes](https://github.com/elastic/kibana/pull/169548)?
> 
> I think it could look better but it is functional..

I think it's a great concept tbh. However, having two different buttons that say "save" is weird. Maybe that pattern can be brought up to our designer to see if they want to corporate that into the new timeline designs.

janmonschke committed:
Merge branch 'main' into save-timeline-improvements-2

janmonschke committed:
Add tour to announce timeline saving changes

nastasha-solomon commented on x-pack/plugins/security_solution/public/timelines/components/flyout/header/translations.ts:
```diff
@@ -77,3 +81,17 @@ export const TIMELINE_TOGGLE_BUTTON_ARIA_LABEL = ({
+export const SAVE_TOUR_TITLE = i18n.translate(
+  'xpack.securitySolution.timeline.flyout.saveTour.title',
+  {
+    defaultMessage: 'Timelines are now saved manually',
```
@janmonschke are we saying that users have to manually save Timelines now? If so, I'd change this text to the following:

```suggestion
    defaultMessage: 'Timelines now require manual saves',
```

nastasha-solomon commented on x-pack/plugins/security_solution/public/timelines/components/flyout/header/index.tsx:
```diff
@@ -451,6 +477,40 @@ const FlyoutHeaderComponent: React.FC<FlyoutHeaderProps> = ({ timelineId }) => {
+                  <EuiText>
+                    <FormattedMessage
+                      id="xpack.securitySolution.timeline.flyout.saveTour.description"
+                      defaultMessage="Use the {saveButton} button to save the timeline."
```
```suggestion
                      defaultMessage="Click {saveButton} to manually save Timeline changes."
```

nastasha-solomon reviewed (approved):
Left two minor suggestions. Hope they help!

janmonschke committed:
Update copy

Co-authored-by: Nastasha Solomon <79124755+nastasha-solomon@users.noreply.github.com>

janmonschke committed:
Update copy

Co-authored-by: Nastasha Solomon <79124755+nastasha-solomon@users.noreply.github.com>

janmonschke committed:
move tour to SaveTimelineButton

janmonschke committed:
introduce `disableTimelineSaveTour` flag to fix cypress tests

janmonschke committed:
disable timeline save tour in ftr tests as well

janmonschke committed:
Revert "disable timeline save tour in ftr tests as well"

This reverts commit e7c44f5934ec4ae73ac2228dc8684b6b0b3f0cb3.

janmonschke commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into save-timeline-improvements-2

janmonschke committed:
disable timeline save tour in ftr tests

janmonschke committed:
fix config typo

janmonschke committed:
disable tour in more tests

ashokaditya reviewed (approved):
`test/security_solution_endpoint` changes LGTM.

MadameSheema commented on x-pack/test/security_solution_cypress/cypress/tasks/timeline.ts:
```diff
@@ -344,6 +346,29 @@ export const expandFirstTimelineEventDetails = () => {
+export const saveTimeline = () => {
+  cy.get(TIMELINE_SAVE_MODAL_OPEN_BUTTON).first().click();
+
+  cy.get('[data-test-subj="save-timeline-modal"]').within(() => {
```
NIT: Please, wrap all the selectors in meaningful screen constants.

janmonschke replied:
Good point! Did it here: https://github.com/elastic/kibana/pull/169239/commits/320fd676e41fb962b8b488c1e51f777af85abdec

MadameSheema commented on x-pack/test/security_solution_cypress/cypress/tasks/timeline.ts:
```diff
@@ -344,6 +346,29 @@ export const expandFirstTimelineEventDetails = () => {
+    cy.get(TIMELINE_TITLE_INPUT)
+      .invoke('val')
+      .then((value) => {
+        if (!value) {
```
We should avoid having conditionals inside tasks. What is the reason for adding this one?

janmonschke replied:
@MadameSheema If no title has been set for the timeline, we would not be able to save it. This just makes sure to set a title before saving. It allows us to use `saveTimeline()` in our tests without passing a title attribute. At the same time, this code makes sure not to overwrite a previous title so that all previous assertions still work.

I can add a comment here, explaining why there is a conditional.

janmonschke replied:
@MadameSheema As agreed, I cleaned up the task here: https://github.com/elastic/kibana/pull/169239/commits/062915a65b0ad8a44afc095f60c5bec9f543aaf2

gergoabraham commented on x-pack/test/security_solution_endpoint/endpoint.config.ts:
```diff
@@ -24,5 +24,10 @@ export default async function (ftrConfigProviderContext: FtrConfigProviderContex
     junitReportName: 'X-Pack Endpoint Functional Tests on ESS',
     target: 'ess',
     services,
+    kbnServerArgs: [
+      `--xpack.securitySolution.offeringSettings=${JSON.stringify({
+        disableTimelineSaveTour: true,
+      })}`,
+    ],
```
hey @janmonschke,

could you please remove the newly added `offeringSettings` flag from `endpoint.config.ts` and from `serverless.endpoint.config.ts`, and instead add it to `x-pack/test/security_solution_endpoint/config.base.ts`? that is the common config for these FTRs - I believe that would be a better place for this non environment specific flag.

janmonschke replied:
@gergoabraham Good point, I did that here: https://github.com/elastic/kibana/pull/169239/commits/8b496838e1a295f70f1cf1ef5fd09846fa6d792a

gergoabraham replied:
awesome, thank you! 🙇

janmonschke committed:
move flag to shared config

janmonschke committed:
use selector constants

gergoabraham reviewed (approved):
@elastic/security-defend-workflows changes look great!

semd commented on x-pack/test/security_solution_cypress/config.ts:
```diff
@@ -47,6 +47,9 @@ export default async function ({ readConfigFile }: FtrConfigProviderContext) {
         `--xpack.securitySolution.enableExperimental=${JSON.stringify([
           'chartEmbeddablesEnabled',
         ])}`,
+        `--xpack.securitySolution.offeringSettings=${JSON.stringify({
+          disableTimelineSaveTour: true,
+        })}`,
```
The goal of the `offeringSettings` flags is to allow us to enable/disable different features in serverless vs ESS "offerings". This new flag has been created only to disable a feature for cypress tests, which is not a correct use of this tool.

If it is temporary I am fine keeping this `disableTimelineSaveTour` flag here because it solves the problem, but after the tests are adapted we will have to clean it and remove the flag. 

If this was planned to be permanent I would suggest finding another way to disable the tour for cypress. Maybe we could set the localStorage key that is already being used, or just add a task to click the finish button in the tests. 

wdyt?

janmonschke replied:
@semd I am using `offeringSettings` to disable the tour to follow a previous pattern that is used in Kibana with `disableWelcomeScreen`. That one is also only used for disabling a tour in tests.

I agree though that pre-populating `localStorage` would be the better way but I didn't manage to make that work at this point. The config flag is the solution that made it work in cypress and ftr tests.

janmonschke replied:
@semd However, I agree that the implied function of the name `offeringSettings` and what the `disableWelcomeScreen` flag do, do not overlap. I could not find an easier way to get access to config variables that are passed to Kibana. Are you aware of another way to get this flag?

semd replied:
`disableWelcomeScreen` flag is set in a direct config of the `home` plugin, we could set a new config in a similar way for the security_solution plugin, or we could also just use `experimentalFeatures`, which has a more generic scope. Let me pull your branch and try a couple of things.

janmonschke committed:
Cypress test improvements

- Make saveTimeline not set the title
- Use addNameToTimeline where a title needs to be set
- Rename addNameToTimeline to addNameToTimelineAndSave

janmonschke commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into save-timeline-improvements-2

logeekal reviewed (approved):
LGTM 🚀 . Thanks for adding the tour and implementing this functionality.

michaelolo24 reviewed (approved):
Awesome work! Thank you for getting this done. 🚀 LGTM!

semd reviewed (changes requested):
About the new `disableTimelineSaveTour` flag in `offeringSettings`. This config should contain flags to enable/disable features in serverless, the new flag is not being used for this purpose, it is used only for testing purposes, nothing to do with offerings. 
I think we should maintain the original purpose of the `offeringSettings` otherwise it will become another kind of experimentalFeatures. So, this should be done in some different way. My suggestion is:

a) Do not disable the tour this way, let tests run normally, if some test fails because the tour popover overlaps with some element, close it with a reusable task like:

```
export const closeTimelineSaveTour = () => {
  if (cy.get('[data-test-subj="timeline-save-tour-close-button"]')) {
    cy.get('[data-test-subj="timeline-save-tour-close-button"]').click();
  }
};
```

b) Use `experimentalFeatures` the same way we are currently using `offeringSettings`, or create a new config (`testingSettings`)  similar to `offeringSettings`.

I've tried localStorage and you are right I couldn't make it work, not sure if we are missing something.

janmonschke committed:
move flag from offeringSettings to enableExperimental

janmonschke commented:
@semd I move the flag to `enableExperimental` here: https://github.com/elastic/kibana/pull/169239/commits/aff46278a7c82aa9d2ba7081e8d639b5072dd7d7. 

I was considering adding a custom cypress task for it before as well but it prove to be more of a pain. The tour component sits on top of EVERY other component, so it was breaking way too many tests :D Adding a task call to all of those tests just didn't seem like a good idea.

janmonschke committed:
Merge branch 'main' into save-timeline-improvements-2

semd reviewed (approved):
Thanks for addressing the comment! LGTM!

semd commented:
@janmonschke 
> The tour component sits on top of EVERY other component, so it was breaking way too many tests

Oh, I ran some cypress tests and they were able to add filters, save the timeline, and do almost everything with the tour popover open. It was only covering the Data View selector, so I thought only those tests needed to close the tour. I might be missing something. 

Anyway. Great work! 👍

MadameSheema reviewed (approved):
security-engineering-productivity changes LGTM! :)
